commit 0e08ae2844409e030cb2a949e077f805f561a146
Author: LeeDrOiD <leedroid.lb@gmail.com>
Date:   Mon May 16 12:13:13 2011 +0100

    drivers/cpufreq/cpufreq_stats.c Prevent memory leak on hotplug (Pershoot)

diff --git a/drivers/cpufreq/cpufreq_stats.c b/drivers/cpufreq/cpufreq_stats.c
index 00d73fc..b8574e8 100644
--- a/drivers/cpufreq/cpufreq_stats.c
+++ b/drivers/cpufreq/cpufreq_stats.c
@@ -305,6 +305,26 @@ static int cpufreq_stat_notifier_trans(struct notifier_block *nb,
 	return 0;
 }
 
+static int cpufreq_stats_create_table_cpu(unsigned int cpu)
+{
+  struct cpufreq_policy *policy;
+  struct cpufreq_frequency_table *table;
+  int ret = -ENODEV;
+  policy = cpufreq_cpu_get(cpu);
+  if (!policy)
+    return -ENODEV;
+  
+  table = cpufreq_frequency_get_table(cpu);
+  if (!table)
+    goto out;
+  
+  ret = cpufreq_stats_create_table(policy, table);
+  
+out:
+    cpufreq_cpu_put(policy);
+    return ret;
+}
+
 static int __cpuinit cpufreq_stat_cpu_callback(struct notifier_block *nfb,
 					       unsigned long action,
 					       void *hcpu)
@@ -316,10 +336,13 @@ static int __cpuinit cpufreq_stat_cpu_callback(struct notifier_block *nfb,
 	case CPU_ONLINE_FROZEN:
 		cpufreq_update_policy(cpu);
 		break;
-	case CPU_DEAD:
-	case CPU_DEAD_FROZEN:
+	case CPU_DOWN_PREPARE:
+	case CPU_DOWN_PREPARE_FROZEN:
 		cpufreq_stats_free_table(cpu);
 		break;
+	case CPU_DOWN_FAILED:
+	case CPU_DOWN_FAILED_FROZEN:
+	cpufreq_stats_create_table_cpu(cpu);
 	}
 	return NOTIFY_OK;
 }
@@ -327,6 +350,7 @@ static int __cpuinit cpufreq_stat_cpu_callback(struct notifier_block *nfb,
 static struct notifier_block cpufreq_stat_cpu_notifier __refdata =
 {
 	.notifier_call = cpufreq_stat_cpu_callback,
+	.priority = 1
 };
 
 static struct notifier_block notifier_policy_block = {
